### PR TITLE
AggregatePart and AggregateRoot improvements.

### PR DESCRIPTION
**Pull request by IlliaShepilov**
Summary of changes:

- Added `getPartState(Class<S>)` method into `AggregatePart`.
- Improved `AggregatePartBuilder`.
- `AggregateRoot` is created and is put into according `AggregatePart` by framework now.
- Created methods in the `AbstractEntity` class to obtain entities by specified constructor parameters.  

Now to get state of the part from other `AggregatePart` we must not create `AggregateRoot` and pass to `AggregateRoot` constructor the `BoundedContext` instance. For now just need to call `getPartState` method and pass the class of the desirable `AggregatePart` state.

Before:
```java
aggregatePartMethod(...){
       final BoundedContext boundedContext = ... (obtain bounded context from somewhere somehow)
       final AggregateRoot root = new AggregateRoot(boundedContext, id);
       final OtherAggregatePartState state = root.getPartState(OtherAggregatePartState.class);
}
```

Now:
```java
aggregatePartMethod(...){
     final OtherAggregatePartState state = getPartState(OtherAggregatePartState.class);}
}
```

### CODE REVIEW

**Comment by IlliaShepilov** [comment]
Summary of changes:

- Added `getPartState(Class<S>)` method into `AggregatePart`.
- Improved `AggregatePartBuilder`.
- Created methods in the `AbstractEntity` class to obtain entities by specified constructor parameters.

**Comment by IlliaShepilov** [comment]
@armiol PTAL.

**Comment by IlliaShepilov** [comment]
@armiol PTAL.

File: server/src/main/java/org/spine3/server/entity/AbstractEntity.java [hunk 1]
```diff
@@ -157,12 +160,126 @@ private static IllegalStateException noSuchConstructor(String entityClass, Strin
...
+     * <p>The part class must have a constructor with ID and {@code AggregateRoot} parameters.
+     *
+     * @param entityClass the {@code AggregatePart} class
+     * @param idClass     the class of entity identifiers
+     * @param <E>         the entity type
+     * @param <I>         the ID type
+     * @return the {@code AggregatePart} constructor
+     * @throws IllegalStateException if the entity class does not have the required constructor
+     */
+    public static <E extends AggregatePart, I, R extends AggregateRoot<I>> Constructor<E>
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/entity/AbstractEntity.java]: It isn't clear why it is `public' and `static`. I though that should be hidden from the end-users.

File: server/src/main/java/org/spine3/server/entity/AbstractEntity.java [hunk 2]
```diff
@@ -157,12 +160,126 @@ private static IllegalStateException noSuchConstructor(String entityClass, Strin
...
+     * @param entityClass the {@code AggregatePart} class
+     * @param idClass     the ID class
+     * @param <E>         the entity type
+     * @param <I>         the ID type
+     * @return the obtained constructor, otherwise {@code null}
+     */
+    @Nullable
+    @SuppressWarnings("unchecked")
+    // It is OK because the constructor arguments are checked, before returning the constructor.
+    private static <E extends AggregatePart, I> Constructor<E> getAggregatePartSupertypeCtr
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/entity/AbstractEntity.java]: `Ctor`.

Comment (armiol) [inline server/src/main/java/org/spine3/server/entity/AbstractEntity.java]: These `static` reflection-based methods seem to be a part of an external utility class. Please consider that to make the `AbstractEntity` code cleaner.

File: server/src/main/java/org/spine3/server/entity/AbstractEntity.java [hunk 3]
```diff
@@ -157,12 +160,126 @@ private static IllegalStateException noSuchConstructor(String entityClass, Strin
     }
 
     /**
+     * Obtains constructor for the passed aggregate part class.
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/entity/AbstractEntity.java]: An article before `constructor` is missing.

Also I would additionally explain which constructor is expected, as long as an end-user is able to define a custom one in the descendant classes.

File: server/src/main/java/org/spine3/server/aggregate/AggregatePart.java [hunk 4]
```diff
@@ -50,10 +50,26 @@
...
+    }
+
+    /**
+     * Obtains a part state by its class.
+     *
+     * @param partStateClass the class of the state of the part
+     * @param <P>            the type of the part state
+     * @return the state of the part or a default state if the state was not found
+     */
+    protected <P extends Message, A extends AggregatePart<I, P, ?>> P getPartState(
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/aggregate/AggregatePart.java]: Do you expect any exceptions in case there is no a part with the given class, i.e. `Class<P>`? I think you should.

**Comment by armiol** [comment]
@IlliaShepilov also please expand the description of the PR by including a simple sample code. That would help a lot when we compose the release notes.

File: server/src/main/java/org/spine3/server/entity/AbstractEntity.java (hunk 1 above)
Comment (IlliaShepilov) [inline server/src/main/java/org/spine3/server/entity/AbstractEntity.java]: I think it is similar case with `getConstructor(Class<E> entityClass, Class<I> idClass)` method. And it is used outside the package in the `EntityBuilder` class and in the `AggregateRepository` class. If make it non-public and non-static we cant use it in the `testutil` module. And then it should be moved to the classes where it is used. Method is written by me is used in the `AggregatePartRepository` and `AggregatePartBuilder` classes.

File: server/src/main/java/org/spine3/server/aggregate/AggregatePart.java (hunk 4 above)
Comment (IlliaShepilov) [inline server/src/main/java/org/spine3/server/aggregate/AggregatePart.java]: I expect `IllegalStateException` in that case, but it mentioned in the documentation for the `getRepository` method in the `AggregateRoot` class. Should I repeat it in the documentation for the `getPartState` method in the `AggregatePart` class?

File: server/src/main/java/org/spine3/server/entity/AbstractEntity.java (hunk 3 above)
Comment (IlliaShepilov) [inline server/src/main/java/org/spine3/server/entity/AbstractEntity.java]: Article is added. I described it in the next paragraph. Should I expand somehow it?

File: server/src/main/java/org/spine3/server/entity/AbstractEntity.java (hunk 1 above)
Comment (IlliaShepilov) [inline server/src/main/java/org/spine3/server/entity/AbstractEntity.java]: Moved to the separate class (`Entities`).

**Comment by IlliaShepilov** [comment]
@armiol PTAL.

File: server/src/main/java/org/spine3/server/aggregate/AggregatePart.java (hunk 4 above)
Comment (armiol) [inline server/src/main/java/org/spine3/server/aggregate/AggregatePart.java]: Let's discuss.

File: server/src/main/java/org/spine3/server/entity/Entities.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+     *
+     * @param entityClass the entity class
+     * @param idClass     the class of entity identifiers
+     * @param <E>         the entity type
+     * @param <I>         the ID type
+     * @return the constructor
+     * @throws IllegalStateException if the entity class does not have the required constructor
+     */
+    public static <E extends Entity<I, ?>, I> Constructor<E>
+    getConstructor(Class<E> entityClass, Class<I> idClass) {
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/entity/Entities.java]: Doesn't it fit a single line?

File: server/src/main/java/org/spine3/server/entity/Entities.java [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+        }
+    }
+
+    /**
+     * Obtains the constructor for the passed aggregate part class.
+     *
+     * <p>Returns the constructor if the first argument is aggregate ID
+     * and the second is {@code AggregateRoot}. For example:
+     *
+     * <pre>{@code AggregatePartCtor(AnAggregateId id, AggregateRoot root).}</pre>
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/entity/Entities.java]: Please remove the `.` near the end of this line.

File: server/src/main/java/org/spine3/server/entity/Entities.java [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+     *
+     * <p>Does not return constructor if the second constructor parameter
+     * is subtype of the {@code AggregateRoot}. For example:
+     *
+     * <pre> {@code SubAggregateRoot extends AggregateRoot{...};}
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/entity/Entities.java]: Please have a single multiline `{@code ...}`.

File: server/src/main/java/org/spine3/server/entity/Entities.java [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+
+    /**
+     * Obtains the constructor for the passed aggregate part class.
+     *
+     * <p>Returns the constructor if the first argument is aggregate ID
+     * and the second is {@code AggregateRoot}. For example:
+     *
+     * <pre>{@code AggregatePartCtor(AnAggregateId id, AggregateRoot root).}</pre>
+     *
+     * <p>Does not return constructor if the second constructor parameter
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/entity/Entities.java]: Articles?

Also, it does return the value, but I think it's `{@code null}`.

File: server/src/main/java/org/spine3/server/entity/Entities.java [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+     * is subtype of the {@code AggregateRoot}. For example:
+     *
+     * <pre> {@code SubAggregateRoot extends AggregateRoot{...};}
+     * {@code AggregatePartCtor(AnAggregateId id, SubAggregateRoot root)};</pre>
+     *
+     * @param entityClass the {@code AggregatePart} class
+     * @param idClass     the ID class of the {@code AggregatePart} class
+     * @param <E>         the {@code Message} entity type
+     * @param <I>         the {@code Message} ID type of the {@code idClass}
+     * @return the obtained constructor, if constructor was not found,
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/entity/Entities.java]:
```
@return the constructor for the Aggregate part class, or {@code null} if no appropriate constructor is found.
```

File: server/src/main/java/org/spine3/server/entity/Entities.java [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+     * the {@code null} will be returned
+     */
+    @Nullable
+    @SuppressWarnings("unchecked")
+    // It is OK because the constructor arguments are checked, before returning the constructor.
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/entity/Entities.java]: Remove the comma.

File: server/src/main/java/org/spine3/server/entity/Entities.java [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+    createAggregateRootEntity(I id, BoundedContext boundedContext, Class<R> rootClass) {
+        checkNotNull(id);
+        checkNotNull(boundedContext);
+        checkNotNull(rootClass);
+
+        try {
+            final Constructor<R> rootConstructor =
+                    rootClass.getDeclaredConstructor(boundedContext.getClass(), id.getClass());
+            rootConstructor.setAccessible(true);
+            R root = rootConstructor.newInstance(boundedContext, id);
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/entity/Entities.java]: There seems to be something missing in this line. Please guess and resolve.

File: server/src/main/java/org/spine3/server/entity/Entities.java (hunk 5 above)
Comment (IlliaShepilov) [inline server/src/main/java/org/spine3/server/entity/Entities.java]: No, it does not.

**Comment by IlliaShepilov** [comment]
@armiol PTAL.

File: server/src/main/java/org/spine3/server/aggregate/AggregateRoot.java [hunk 12]
```diff
@@ -79,11 +79,15 @@ protected BoundedContext getBoundedContext() {
      * Obtains a part state by its class.
      *
      * @param partStateClass the class of the state of the part
-     * @param <S> the type of the part state
+     * @param <S>            the type of the part state
      * @return the state of the part or a default state if the state was not found
+     * @throws IllegalStateException if a repository was not found,
+     *                               or the repository does not match expectations
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/aggregate/AggregateRoot.java]: > expectations.

Please let readers know what they are.

That's a common issue, please address the similar cases.

**Comment by IlliaShepilov** [comment]
@armiol PTAL.

File: server/src/main/java/org/spine3/server/aggregate/AggregatePartRepository.java [hunk 13]
```diff
@@ -24,26 +24,32 @@
...
  *
  * @author Alexander Yevsyukov
  */
-public abstract class AggregatePartRepository
-        <I, A extends AggregatePart<I, ?, ?>, R extends AggregateRoot<I>>
+public abstract class AggregatePartRepository<I, A extends AggregatePart<I, ?, ?>>
         extends AggregateRepository<I, A> {
 
-    private final R root;
+    private final Class<AggregateRoot<I>> rootClass;
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/aggregate/AggregatePartRepository.java]: Please have a description for this field. It is important.

Comment (armiol) [inline server/src/main/java/org/spine3/server/aggregate/AggregatePartRepository.java]: Isn't it `Class<? extends AggregateRoot<I>>`?

File: server/src/main/java/org/spine3/server/aggregate/AggregatePartRepository.java [hunk 14]
```diff
@@ -24,26 +24,32 @@
...
-    protected AggregatePartRepository(BoundedContext boundedContext, R root) {
+    protected AggregatePartRepository(BoundedContext boundedContext) {
         super(boundedContext);
-        this.root = root;
+        rootClass = getAggregateRootClass();
+    }
+
+    private <R extends AggregateRoot<I>> Class<R> getAggregateRootClass() {
+        @SuppressWarnings("unchecked")
+        // It is OK because value returned with `getAggregateRootClass` is fully controlled.
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/aggregate/AggregatePartRepository.java]: `the value`.
And also the sentence doesn't seem to explain anything.

File: server/src/main/java/org/spine3/server/aggregate/AggregatePartRepository.java [hunk 15]
```diff
@@ -24,26 +24,32 @@
...
+        final Class<R> rootClass = (Class<R>) Aggregates.getAggregateRootClass(getEntityClass());
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/aggregate/AggregatePartRepository.java]: After some thought: that's not a good design decision. That's a "Munchausen" pulling himself out of the dirt using own hair.

Let's try to figure out the `root` class from something else: generic parameter, `AggregatePartRepository` constructor etc.

File: server/src/main/java/org/spine3/server/aggregate/AggregatePartRepository.java [hunk 16]
```diff
@@ -57,8 +63,8 @@ protected AggregatePartRepository(BoundedContext boundedContext, R root) {
     @SuppressWarnings("MethodDoesntCallSuperMethod") // The call of the super method is not needed.
     @Override
     public A create(I id) {
-        final Constructor<A> entityConstructor =
-                getAggregatePartConstructor(getEntityClass(), root.getClass(), getIdClass());
+        final Constructor<A> entityConstructor = getEntityConstructor();
+        final AggregateRoot<I> root = createAggregateRootEntity(id, getBoundedContext(), rootClass);
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/aggregate/AggregatePartRepository.java]: `AggregateRoot` is not an `Entity`.

File: server/src/main/java/org/spine3/server/aggregate/Aggregates.java [hunk 17]
```diff
@@ -36,44 +35,9 @@
  *
  * @author Illia Shepilov
  */
-public class Entities {
+public class Aggregates {
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/aggregate/Aggregates.java]: Please find the obvious error that you still have after the renaming.

File: server/src/test/java/org/spine3/server/aggregate/AggregatesShould.java [hunk 18]
```diff
@@ -112,6 +104,20 @@ public void return_appropriate_constructor() {
...
+    @Test
+    public void obtain_aggregate_root_class() {
+        final Class<?> rootClass = getAggregateRootClass(AnAggregatePart.class);
+
+        assertNotNull(rootClass);
+        assertEquals(rootClass.getName(), AnAggregateRoot.class.getName());
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void not_obtain_aggregate_root_class_when_part_does_not_have_appropriate_ctor() {
```
Comment (armiol) [inline server/src/test/java/org/spine3/server/aggregate/AggregatesShould.java]: `..._if_part...`

**Comment by IlliaShepilov** [comment]
@armiol PTAL.

File: server/src/main/java/org/spine3/server/aggregate/AggregateRepository.java [hunk 19]
```diff
@@ -104,20 +106,30 @@ protected AggregateRepository(BoundedContext boundedContext) {
...
+     * Obtains constructor for the passed entity class.
+     *
+     * @return the entity constructor
+     */
+    protected Constructor<A> getEntityConstructor() {
+        final Constructor<A> result = getConstructor(entityClass, idClass);
         return result;
     }
 
+    private void initEntityConstructorIfNeeded() {
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/aggregate/AggregateRepository.java]: That's odd. Why don't you build that lazy init into the `getEntityConstructor`?

File: server/src/main/java/org/spine3/server/aggregate/Aggregates.java [hunk 20]
```diff
@@ -0,0 +1,211 @@
...
+        checkNotNull(root);
+
+        try {
+            final E result = ctor.newInstance(id, root);
+            return result;
+        } catch (InvocationTargetException | InstantiationException | IllegalAccessException e) {
+            throw new IllegalStateException(e);
+        }
+    }
+
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/aggregate/Aggregates.java]: Please remove this line.

File: server/src/test/java/org/spine3/server/aggregate/AggregatesShould.java [hunk 21]
```diff
@@ -0,0 +1,151 @@
...
+    @Before
+    public void setUp() {
+        boundedContext = BoundedContext.newBuilder()
+                                       .build();
+        id = newUuid();
+        root = new AnAggregateRoot(boundedContext, id);
+    }
+
+    @Test
+    public void pass_null_pointer_check() throws NoSuchMethodException {
```
Comment (armiol) [inline server/src/test/java/org/spine3/server/aggregate/AggregatesShould.java]: `not_accept_nulls_as_parameter_values`

File: server/src/test/java/org/spine3/server/command/CommandSchedulingShould.java [hunk 22]
```diff
@@ -121,7 +121,7 @@ public void reschedule_commands_from_storage_in_parallel_on_build_if_thread_spaw
 
         // Sleep to ensure the commands have been rescheduled in parallel.
         try {
-            Thread.sleep(100);
+            Thread.sleep(150);
```
Comment (armiol) [inline server/src/test/java/org/spine3/server/command/CommandSchedulingShould.java]: Why is that changed?

Comment (IlliaShepilov) [inline server/src/test/java/org/spine3/server/command/CommandSchedulingShould.java]: Because that test failed yesterday during the check after updating the branch.
https://travis-ci.org/SpineEventEngine/core-java/builds/204632979
Also one more build was failed yesterday, I think it is because our tests are super fast.
https://travis-ci.org/SpineEventEngine/core-java/builds/204663484

**Comment by IlliaShepilov** [comment]
@armiol, @alexander-aleksandrov PTAL.

**Comment by alexander-yevsyukov** [comment]
> agrgegatePartMethod(...)

@IlliaShepilov please fix the spelling in PR description.

File: server/src/main/java/org/spine3/server/aggregate/AggregatePartRepository.java [hunk 23]
```diff
@@ -22,18 +22,32 @@
...
+     * The {@code AggregateRoot} class, which instance will be injected
+     * into the {@code AggregatePart} instance during initialization.
+     */
+    private final Class<R> rootClass;
 
     /**
      * {@inheritDoc}
      */
-    protected AggregatePartRepository(BoundedContext boundedContext) {
+    protected AggregatePartRepository(BoundedContext boundedContext, Class<R> rootClass) {
```
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/aggregate/AggregatePartRepository.java]: See my earlier comment. We probably can get the class of a root from the generic params of the part class.

File: server/src/main/java/org/spine3/server/aggregate/Aggregates.java [hunk 24]
```diff
@@ -0,0 +1,210 @@
...
+     * @param idClass     the ID class of the {@code AggregatePart} class
+     * @param <E>         the {@code Message} entity type
+     * @param <I>         the {@code Message} ID type of the {@code idClass}
+     * @return the constructor for the Aggregate part class,
+     * or {@code null} if no appropriate constructor is found.
+     */
+    @Nullable
+    @SuppressWarnings("unchecked")
+    // It is OK because the constructor arguments are checked before returning the constructor.
+    private static <E extends AggregatePart, I> Constructor<E> getAggregatePartSupertypeCtor
```
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/aggregate/Aggregates.java]: Please have the method name on the new line, not parameters.

File: server/src/main/java/org/spine3/server/aggregate/Aggregates.java [hunk 25]
```diff
@@ -0,0 +1,210 @@
...
+import static java.lang.String.format;
+
+/**
+ * The utility class for working with aggregates.
+ *
+ * @author Illia Shepilov
+ */
+class Aggregates {
+
+    private Aggregates() {
```
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/aggregate/Aggregates.java]: Please have comment inside the empty body. Like:

```java
// Prevent instantiation of this utility class.
```

File: server/src/main/java/org/spine3/server/aggregate/Aggregates.java [hunk 26]
```diff
@@ -0,0 +1,210 @@
...
+    }
+
+    /**
+     * Obtains the constructor for the passed aggregate part class.
+     *
+     * <p>The part class must have a constructor with ID and {@code AggregateRoot} parameters.
```
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/aggregate/Aggregates.java]: Why do we have that many empty lines in the below Javadoc? Can you make it a bit more condensed.

File: server/src/test/java/org/spine3/server/aggregate/AggregatePartRepositoryLookupShould.java [hunk 27]
```diff
@@ -48,13 +49,18 @@
...
+                                      .build();
+        final ProjectRoot root = new ProjectRoot(boundedContext, id);
         boundedContext.register(new ProjectPartRepository(boundedContext));
         boundedContext.register(new TaskAggregateRepository(boundedContext));
     }
 
     @Test
     public void find_a_repository() {
-        AggregatePartRepositoryLookup lookup = createLookup(boundedContext, ProjectId.class, Project.class);
+        AggregatePartRepositoryLookup lookup = createLookup(boundedContext, ProjectId.class,
```
Comment (alexander-yevsyukov) [inline server/src/test/java/org/spine3/server/aggregate/AggregatePartRepositoryLookupShould.java]: If not one line, then each arg in a new line. Please fix it here and in the rest of occasions.

File: server/src/test/java/org/spine3/server/aggregate/AggregateRootShould.java [hunk 28]
```diff
@@ -85,55 +137,91 @@ private ProjectRoot(BoundedContext boundedContext, ProjectId id) {
...
+        private ProjectLifeCycleRepository(BoundedContext boundedContext) {
+            super(boundedContext, ProjectRoot.class);
         }
     }
 
-    private static class ProjectHeaderRepository extends AggregatePartRepository<ProjectId, ProjectHeader> {
+    private static class MockStreamObserver implements StreamObserver<Response> {
+        @Override
+        public void onNext(Response value) {
+
```
Comment (alexander-yevsyukov) [inline server/src/test/java/org/spine3/server/aggregate/AggregateRootShould.java]: Please document empty method bodies or (assuming this is a mock object) just remove empty lines.

**Comment by IlliaShepilov** [comment]
@armiol, @alexander-yevsyukov PTAL.

File: server/src/main/java/org/spine3/server/aggregate/AggregatePartRepository.java [hunk 29]
```diff
@@ -63,15 +64,15 @@ protected AggregatePartRepository(BoundedContext boundedContext, Class<R> rootCl
...
+        final A result = createAggregatePart(entityConstructor, root);
         return result;
     }
 
     @SuppressWarnings("MethodDoesntCallSuperMethod") // The call of the super method is not needed.
     @Override
     protected Constructor<A> getEntityConstructor() {
         final Constructor<A> result =
-                Aggregates.getAggregatePartConstructor(getEntityClass(), rootClass, getIdClass());
+                Aggregates.getAggregatePartConstructor(getEntityClass(), rootClass);
```
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/aggregate/AggregatePartRepository.java]: Can we get the root class as a generic parameter of the aggregate class or, more straightforward, the repository class? If we can, we won't need to accept it as a constructor parameter.

File: server/src/main/java/org/spine3/server/aggregate/Aggregates.java [hunk 30]
```diff
@@ -68,95 +59,39 @@ private Aggregates() {
...
      * @param entityClass the {@code AggregatePart} class
-     * @param idClass     the class of entity identifiers
      * @param <E>         the entity type
      * @param <I>         the ID type
      * @return the {@code AggregatePart} constructor
      * @throws IllegalStateException if the entity class does not have the required constructor
      */
-    static <E extends AggregatePart, I, R extends AggregateRoot<I>> Constructor<E>
-    getAggregatePartConstructor(Class<E> entityClass, Class<R> rootClass, Class<I> idClass) {
+    static <E extends AggregatePart<I, ?, ?, ?>, I> Constructor<E>
```
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/aggregate/Aggregates.java]: It would be better to have `A` instead of `E` for consistency with the class declaration.

File: server/src/main/java/org/spine3/server/aggregate/Aggregates.java [hunk 31]
```diff
@@ -68,95 +59,39 @@ private Aggregates() {
...
+    getAggregatePartConstructor(Class<E> entityClass, Class<?> rootClass) {
```
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/aggregate/Aggregates.java]: The parameter should be named `aggregatePartClass`.

Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/aggregate/Aggregates.java]: We also can obtain the root class as a generic parameter of the passed aggregate part class.

File: server/src/main/java/org/spine3/server/aggregate/Aggregates.java [hunk 32]
```diff
@@ -68,95 +59,39 @@ private Aggregates() {
...
 
         try {
             final Constructor<E> constructor =
-                    entityClass.getDeclaredConstructor(idClass, rootClass);
+                    entityClass.getDeclaredConstructor(rootClass);
             constructor.setAccessible(true);
             return constructor;
         } catch (NoSuchMethodException ignored) {
-            throw noSuchConstructor(entityClass.getName());
+            throw noSuchConstructor(entityClass.getName(), rootClass.getName());
```
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/aggregate/Aggregates.java]: Please test this branch.

File: server/src/main/java/org/spine3/server/aggregate/Aggregates.java [hunk 33]
```diff
@@ -68,95 +59,39 @@ private Aggregates() {
...
         }
     }
 
-    private static IllegalStateException noSuchConstructor(String entityClass) {
+    private static IllegalStateException noSuchConstructor(String entityClass, String rootClass) {
         final String errMsg = format("%s class must declare a constructor " +
-                                     "with ID and AggregateRoot parameters.", entityClass);
+                                     "with %s AggregateRoot parameter.", entityClass, rootClass);
```
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/aggregate/Aggregates.java]: It probably wouldn't read well:

```
MyAggregatePart class must declare a constructor with MyAggregateRoot AggregateRoot parameter.
```

See the duplicating part at the end?

File: server/src/main/java/org/spine3/server/aggregate/AggregateRoot.java [hunk 34]
```diff
@@ -40,7 +40,8 @@
 
     /** The map from a part class to a repository which manages corresponding aggregate part */
     private static final Map<Class<? extends Message>,
-            AggregatePartRepository<?, ?, ?>> partsAccess = Maps.newConcurrentMap();
+                             AggregatePartRepository<?, ?, ?>> partsAccess =
+                                                               Maps.newConcurrentMap();
 
     /** The bounded context to which the aggregate belongs. */
     private final BoundedContext boundedContext;
```
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/aggregate/AggregateRoot.java]: Please make `getBoundedContext()` private too. We want to limit the ability for the end developer to do “dirty tricks” by accessing B.C. directly.

**Comment by IlliaShepilov** [comment]
@alexander-yevsyukov, @armiol PTAL.

File: server/src/main/java/org/spine3/server/aggregate/AggregatePartRepository.java (hunk 23 above)
Comment (armiol) [inline server/src/main/java/org/spine3/server/aggregate/AggregatePartRepository.java]: @IlliaShepilov please react on this by leaving a comment.

File: server/src/main/java/org/spine3/server/aggregate/AggregatePartRepositoryLookup.java [hunk 35]
```diff
@@ -62,19 +63,23 @@ private AggregatePartRepositoryLookup(BoundedContext boundedContext, Class<I> id
...
-        final AggregatePartRepository<I, A> result = checkIdClass((AggregatePartRepository<?, ?>)repo);
+        final AggregatePartRepository<I, A, ?> result =
+                checkIdClass((AggregatePartRepository<?, ?, ?>) repo);
         return result;
     }
 
     @SuppressWarnings("OptionalUsedAsFieldOrParameterType") // as this is the purpose of the method
-    private AggregateRepository<?, ?> checkFound(Optional<? extends AggregateRepository<?, ?>> rawRepo) {
+    private AggregateRepository<?, ?>
+    checkFound(Optional<? extends AggregateRepository<?, ?>> rawRepo) {
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/aggregate/AggregatePartRepositoryLookup.java]: Why this wrapping again?

File: server/src/main/java/org/spine3/server/aggregate/AggregateRoot.java (hunk 34 above)
Comment (armiol) [inline server/src/main/java/org/spine3/server/aggregate/AggregateRoot.java]: @IlliaShepilov let's discuss why it is still there.

File: server/src/main/java/org/spine3/server/aggregate/Aggregates.java (hunk 26 above)
Comment (armiol) [inline server/src/main/java/org/spine3/server/aggregate/Aggregates.java]: @alexander-yevsyukov that's my idea — in order to make the example code more readable. It's hard to tell what's going on when there are no blank lines.

**Comment by IlliaShepilov** [comment]
@alexander-yevsyukov PTAL.

**Comment by armiol** [comment]
@IlliaShepilov LGTM.

File: server/src/main/java/org/spine3/server/aggregate/AggregatePartRepositoryLookup.java [hunk 36]
```diff
@@ -62,19 +63,23 @@ private AggregatePartRepositoryLookup(BoundedContext boundedContext, Class<I> id
...
         if (!rawRepo.isPresent()) {
-            final String errMsg = String.format("Unable to find repository for the state class: %s", stateClass);
+            final String errMsg = String.format("Unable to find repository for the state class: %s",
```
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/aggregate/AggregatePartRepositoryLookup.java]: Use static import for `String.format()` and it would make the line shorter. Hence you'd avoid wrapping in this case.

File: server/src/main/java/org/spine3/server/aggregate/AggregateRepository.java [hunk 37]
```diff
@@ -109,20 +111,26 @@ protected AggregateRepository(BoundedContext boundedContext) {
...
+     * Obtains constructor for the passed entity class.
+     *
+     * @return the entity constructor
+     */
+    protected Constructor<A> getEntityConstructor() {
+        if (this.entityConstructor == null) {
+            final Constructor<A> result = getConstructor(entityClass, idClass);
+            return result;
+        }
+        return this.entityConstructor;
```
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/aggregate/AggregateRepository.java]: Please test this branch.

Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/aggregate/AggregateRepository.java]: It looks like we need to remember the `result` in `entityConstructor`.

File: server/src/test/java/org/spine3/server/aggregate/AggregateRootShould.java [hunk 38]
```diff
@@ -21,51 +21,71 @@
...
 import static org.mockito.Mockito.atMost;
 import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.verify;
+import static org.spine3.base.Identifiers.newUuid;
+import static org.spine3.testdata.TestCommandContextFactory.createCommandContext;
 
 public class AggregateRootShould {
 
-    private AggregateRoot aggregateRoot;
+    private static final StreamObserver<Response> RESPONSE_OBSERVER = new MockStreamObserver();
```
Comment (alexander-yevsyukov) [inline server/src/test/java/org/spine3/server/aggregate/AggregateRootShould.java]: > RESPONSE_OBSERVER
Please read [Constant Names](https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names) section in Google Java Code Style.

Fix the variable name name accordingly.

File: server/src/test/java/org/spine3/server/aggregate/AggregatesShould.java [hunk 39]
```diff
@@ -0,0 +1,131 @@
...
+        root = new AnAggregateRoot(boundedContext, id);
+    }
+
+    @Test
+    public void not_accept_nulls_as_parameter_values() throws NoSuchMethodException {
+        final Constructor constructor = root.getClass()
+                                            .getDeclaredConstructor(BoundedContext.class,
+                                                                    String.class);
+        final NullPointerTester tester = new NullPointerTester();
+        tester.setDefault(Constructor.class, constructor);
```
Comment (alexander-yevsyukov) [inline server/src/test/java/org/spine3/server/aggregate/AggregatesShould.java]: Why don't use chained calls?

File: server/src/test/java/org/spine3/server/aggregate/AggregatesShould.java [hunk 40]
```diff
@@ -0,0 +1,131 @@
...
+        tester.setDefault(BoundedContext.class, boundedContext);
+        tester.setDefault(AggregateRoot.class, root);
+
+        tester.testStaticMethods(Aggregates.class, NullPointerTester.Visibility.PUBLIC);
```
Comment (alexander-yevsyukov) [inline server/src/test/java/org/spine3/server/aggregate/AggregatesShould.java]: There's `testAllStaticMethods()`, please use it. It would read better.

Comment (IlliaShepilov) [inline server/src/test/java/org/spine3/server/aggregate/AggregatesShould.java]: There is `testAllPublicStaticMethods()`. The `Aggregates` class has methods with `default` visibility level.

File: server/src/test/java/org/spine3/server/aggregate/AggregateRootShould.java (hunk 38 above)
Comment (IlliaShepilov) [inline server/src/test/java/org/spine3/server/aggregate/AggregateRootShould.java]: But it is mock object and whose methods have no detectable side effects. They do nothing:
```java
    private static class MockStreamObserver implements StreamObserver<Response> {
        @Override
        public void onNext(Response value) {
        }

        @Override
        public void onError(Throwable t) {
        }

        @Override
        public void onCompleted() {
        }
    }
```
The class is not marked as `final`, it is because of that it cannot be declared as constant?

File: server/src/test/java/org/spine3/server/aggregate/AggregatesShould.java (hunk 40 above)
Comment (IlliaShepilov) [inline server/src/test/java/org/spine3/server/aggregate/AggregatesShould.java]: Sorry, that test has incorrect behavior, it should test all methods starting from the `default` visibility level. It is fixed now.

**Comment by IlliaShepilov** [comment]
@alexander-yevsyukov PTAL.

File: server/src/test/java/org/spine3/server/aggregate/AggregateRootShould.java (hunk 38 above)
Comment (alexander-yevsyukov) [inline server/src/test/java/org/spine3/server/aggregate/AggregateRootShould.java]: This is not a constant by definition. You do have methods that are meant to mutate the state. 

There's also `Tests.emptyObserver()` created specifically for this.

Comment (IlliaShepilov) [inline server/src/test/java/org/spine3/server/aggregate/AggregateRootShould.java]: Variable is renamed.
Mock class is replaced with `emptyObserver()`.
